src/palimpzest/sets.py

```diff
@@ -129,52 +108,33 @@ def json_schema(self):
 
 class Dataset(Set):
     """
-    A Dataset is the intended abstraction for programmers to interact with when manipulating Sets.
-
-    Users instantiate a Dataset by specifying a `source` that either points to a
-    DataSource or an existing cached Set. Users can then perform computations on
-    the Dataset in an imperative fashion by leveraging functions such as `filter`,
-    `convert`, `aggregate`, etc. Underneath the hood, each of these operations creates
-    a new Set which is cached by the DataManager. As a result, the Sets define the
-    lineage of computation on a Dataset, and this enables programmers to re-use
-    previously cached computation by providing it as a `source` to some future Dataset.
-    """
+    A Dataset is the intended abstraction for programmers to interact with when writing PZ programs.
 
-    def __init__(self, source: str | list | pd.DataFrame | DataSource, schema: Schema | None = None, *args, **kwargs):
-        # convert source (str) -> source (DataSource) if need be
-        updated_source = DataDirectory().get_or_register_dataset(source) if isinstance(source, (str, list, pd.DataFrame)) else source
+    Users instantiate a Dataset by specifying a `source` that either points to a DataReader
+    or an existing Dataset. Users can then perform computations on the Dataset in a lazy fashion
+    by leveraging functions such as `filter`, `sem_filter`, `sem_add_columns`, `aggregate`, etc.
+    Underneath the hood, each of these operations creates a new Dataset. As a result, the Dataset
+    defines a lineage of computation.
+    """
 
-        if schema is None:
-            # This is mainly for DataSource with a schema. 
-            if updated_source.schema is not None:
-                schema = updated_source.schema
-            else:
-                schema = Schema.from_df(source) if isinstance(source, pd.DataFrame) else DefaultSchema
+    def __init__(
+        self,
+        source: str | Path | list | pd.DataFrame | DataReader | Dataset,
+        schema: Schema | None = None,
+        *args,
+        **kwargs,
+    ) -> None:
+        # NOTE: this function currently assumes that DataReader will always be provided with a schema;
```

Comment:
I am not a big fan of how this is implemented now. Maybe, we can fix this in a later PR. But I think we should distinguish between a parent and a reader. 

A dataset should have a member reader. This constructor should take in an argument reader: str | Path | list | pd.DataFrame | DataReader and set the reader member. 

Constructor of the class Set should take an argument parent which should be of type Set | None. It should be set to None here (the first dataset in a query dag). 

This is cleaner and you can avoid some of the isinstance(node/source, Datareader) calls here and in your optimizer.